### PR TITLE
feat: kpi

### LINKED ISSUE #102: Change gap chart to dose information when group selected
The gap chart should only be available when users have "all residents" selected in the drop-down. Otherwise, it should show a KPI dashboard-like view with dose information on the selected group.

### PR DESCRIPTION
**Pull request by eldu**
- Adds KPI in place of gap chart for when a focus stat is selected

## Demo
Screen Shot 2022-05-03 at 11 41 36 AM
Screen Shot 2022-05-03 at 11 41 27 AM
Screen Shot 2022-04-29 at 3 51 48 PM
Screen Shot 2022-04-29 at 3 44 06 PM

## Old Images
[100% Vaccinated](https://user-images.githubusercontent.com/5270855/166060302-d2c7b249-5574-4664-9248-15a5cb89c1a8.png)
[Not enough information](https://user-images.githubusercontent.com/5270855/166059225-46df8962-17aa-432f-8923-1c4f3a9fdb12.png)

closes #102, #75

### CODE REVIEW

File: src/components/dashboard/GapByRace.vue [hunk 1]
```diff
@@ -1,26 +1,74 @@
 <template>
-  <div class="gap-container">
-    <div>
-      <GapChart
-        :active-stats="activeStats"
-        :expected="expected"
-        :field-data="fieldData"
-      />
+  <div class="outer-container">
+    <div
```
Comment (mcmcgrath13) [inline src/components/dashboard/GapByRace.vue:3]: maybe add a comment above each of the three main chunks with what is happening

File: src/components/dashboard/GapByRace.vue [hunk 2]
```diff
@@ -1,26 +1,74 @@
...
+        visible:
+          focusStat.value !== 'total' && activeFocusStats?.population > 0,
+      }"
+    >
+      <div
+        class="is-flex is-flex-direction-row is-justify-content-space-around"
+      >
+        <KeyPerformanceIndicator
+          :value="formatPct(activeFocusStats?.pct)"
+          title="Vaccinated in Community"
```
Comment (mcmcgrath13) [inline src/components/dashboard/GapByRace.vue:42]: should this title more clearly specify the race selected? @s-bessey

File: src/components/dashboard/GapByRace.vue [hunk 3]
```diff
@@ -1,26 +1,74 @@
...
+      <div class="centered">
+        <p class="has-text-centered">
+          In {{ activeCluster.name }},
+          <strong>{{ formatPct(activeFocusStats?.pct) }}</strong> of
+          {{ activeFocusStats?.name }} residents are are vaccinated compared to
+          our goal of {{ formatPct(expected) }} total vaccinations statewide.
+          <strong
+            >{{ activeFocusStats?.gap }} more
+            {{ activeFocusStats?.name }} residents</strong
+          >
```
Comment (mcmcgrath13) [inline src/components/dashboard/GapByRace.vue:58]: if the gap is 0, should this text change?

File: src/components/dashboard/GapByRace.vue [hunk 4]
```diff
@@ -29,12 +77,14 @@
 import { computed } from "vue";
 
 import GapChart from "@/components/dashboard/GapChart.vue";
+import KeyPerformanceIndicator from "./KeyPerformanceIndicator.vue";
```
Comment (mcmcgrath13) [inline src/components/dashboard/GapByRace.vue:80]: use the `@/components/...` style instead for consistency

File: src/components/dashboard/GapByRace.vue [hunk 5]
```diff
@@ -86,16 +142,33 @@ const minVaxRace = computed(() => {
...
+.outer-container {
+  display: grid;
+  grid-template: 1fr / 1fr;
+  place-items: center;
+
+  > * {
+    grid-column: 1 / 1;
+    grid-row: 1 / 1;
+  }
+}
```
Comment (mcmcgrath13) [inline src/components/dashboard/GapByRace.vue:173]: what does this do? I'd think this would be a use case for absolute positioning?

File: src/components/dashboard/KeyPerformanceIndicator.vue [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+<script lang="ts" setup>
+defineProps<{
+  title: string;
+  value: number | string;
+}>();
+</script>
+
+<style scoped>
+.kpi-container {
+  gap: 1.5rem;
```
Comment (mcmcgrath13) [inline src/components/dashboard/KeyPerformanceIndicator.vue:23]: it looks weird to me to have a gap on this container when it's not obviously in a gap or flex - should some styles move around so this looks more typical?

**Review by maxwellkrieger — CHANGES_REQUESTED** [review]
First of all, I am LOVING the look of the KPI!! 😍 😍 😍 
Only change I would make is to slightly modify the KPI text (unless it ends up looking too wordy):

9% of [selectedgroup] residents are vaccinated in this community
214 vaccine doses for [selectedgroup] residents are needed to close the gap

**Review by s-bessey — COMMENTED** [review]
Looking good!

File: src/components/dashboard/GapByRace.vue (hunk 5 above)
Comment (eldu) [reply src/components/dashboard/GapByRace.vue:173]: This stacks all the child divs on top of each other while still taking up space. So the size of the container ends up being the max size of the children.

Here's the stack overflow. I'll add it in a comment on top of this piece of code
https://stackoverflow.com/questions/1909648/stacking-divs-on-top-of-each-other

Comment (eldu) [reply src/components/dashboard/GapByRace.vue:173]: Absolute elements don't take up space in a container though, so sizing that outer container with absolute elements is confusing?

File: src/components/dashboard/GapByRace.vue (hunk 3 above)
Comment (tdivoll) [reply src/components/dashboard/GapByRace.vue:58]: agreed, in the case below, maybe the text should report the stats on the Asian gap and/or state that there is not enough data on vaccination rates for Black residents of the selected area.

File: src/components/dashboard/GapByRace.vue (hunk 1 above)
Comment (eldu) [reply src/components/dashboard/GapByRace.vue:13]: I added in comments! let me know if there is anything i need to elaborate on !

File: src/components/dashboard/KeyPerformanceIndicator.vue (hunk 6 above)
Comment (eldu) [reply src/components/dashboard/KeyPerformanceIndicator.vue]: I updated this style !

File: src/components/dashboard/GapByRace.vue (hunk 4 above)
Comment (eldu) [reply src/components/dashboard/GapByRace.vue]: Sounds good!

File: src/components/dashboard/GapByRace.vue (hunk 3 above)
Comment (eldu) [reply src/components/dashboard/GapByRace.vue:70]: I've updated this wording !

File: src/components/dashboard/GapByRace.vue (hunk 2 above)
Comment (eldu) [reply src/components/dashboard/GapByRace.vue]: I've updated this wording !

**Review by tdivoll — COMMENTED** [review]
The KPI for a selected group looks great! Suggested catching the cases where all residents are selected and the text states 0% vacc, 0 needed, but it's due to a lack of data. Also love the map update where groups with lack of data have the diagonal stripe shading.

File: src/components/dashboard/GapByRace.vue (hunk 5 above)
Comment (mcmcgrath13) [reply src/components/dashboard/GapByRace.vue:188]: right right.  I wonder if this would be more clear if you used a named template then assigned every child's grid area to that named area?

File: src/components/dashboard/GapByRace.vue (hunk 3 above)
Comment (eldu) [reply src/components/dashboard/GapByRace.vue:70]: Hmm, working on this !

Comment (eldu) [reply src/components/dashboard/GapByRace.vue:75]: Screen Shot 2022-05-03 at 10 46 33 AM
 Updated to this ~

File: src/components/dashboard/GapByRace.vue (hunk 5 above)
Comment (eldu) [reply src/components/dashboard/GapByRace.vue]: Style updated ! Thanks for the help on this Mary !

**Comment by eldu** [comment]
> The KPI for a selected group looks great! Suggested catching the cases where all residents are selected and the text states 0% vacc, 0 needed, but it's due to a lack of data. Also love the map update where groups with lack of data have the diagonal stripe shading.

We should cover all the cases! Though, the case you mention should never be hit. I have adjusted my code such that now it's simply an if/else. if the focus stat is all residents or not. So we should be good on this front now !

File: src/components/dashboard/GapByRace.vue [hunk 7]
```diff
@@ -1,26 +1,110 @@
...
-      <GapChart
-        :active-stats="activeStats"
-        :expected="expected"
-        :field-data="fieldData"
-      />
+  <!--
+    Depending on the controls, display different charts and layouts
+
+    Implementation:
+    Each child of the outer-container is stacked on top of each other taking up space and his hidden based on the
```
Comment (mcmcgrath13) [inline src/components/dashboard/GapByRace.vue:6]:
```suggestion
    Each child of the outer-container is stacked on top of each other taking up space and is hidden based on the
```

new lines could maybe use reformatting as well, looks a bit overflowy

File: src/components/dashboard/GapByRace.vue [hunk 8]
```diff
@@ -1,26 +1,110 @@
...
+        <GapChart
+          :active-stats="activeStats"
+          :expected="expected"
+          :field-data="fieldData"
+        />
+      </div>
+      <div class="centered">
+        <p class="has-text-centered">
+          In {{ activeCluster.name }}, the largest gap is among
+          {{ activeStats[0]?.name }} residents. Only
```
Comment (mcmcgrath13) [inline src/components/dashboard/GapByRace.vue:27]: this used to be `minVaxRace`, why no longer? is the first one guaranteed to be the gappiest?

File: src/components/dashboard/GapByRace.vue [hunk 9]
```diff
@@ -1,26 +1,110 @@
...
+          {{ activeFocusStats?.name }} residents are vaccinated compared to our
+          goal of {{ formatPct(expected) }} total vaccinations statewide.
+          <strong
+            >{{ activeFocusStats?.gap }} more
+            {{ activeFocusStats?.name }} residents</strong
+          >
+          need to be vaccinated to close this gap.
+        </p>
+        <!-- There is no gap in this focus group, display the largest gap -->
+        <span v-else class="has-text-centered">
```
Comment (mcmcgrath13) [inline src/components/dashboard/GapByRace.vue:79]: should this class be moved to the outer container?

File: src/components/dashboard/GapByRace.vue [hunk 10]
```diff
@@ -58,44 +144,59 @@ const activeStats = computed(() => {
...
+          name: f.name,
+          pct:
+            population > 0
+              ? Math.min(1, row[f.observedField] / row[f.populationField])
+              : NaN,
+          gap: Math.max(0, row[f.expectedField] - row[f.observedField]),
+          population,
+        };
+      })
+      .sort(sortByProperty("pct"));
```
Comment (mcmcgrath13) [inline src/components/dashboard/GapByRace.vue:160]: ah, I see it is sorted

instead of referring to the 0th item everywhere, I still think I'd pull that one out to a computed value that takes the 0th item and is named more meaningfully

File: src/components/dashboard/KeyPerformanceIndicator.vue [hunk 11]
```diff
@@ -0,0 +1,30 @@
+<template>
+  <div class="kpi-container mx-3">
+    <div class="centered">
```
Comment (mcmcgrath13) [inline src/components/dashboard/KeyPerformanceIndicator.vue:3]: can these div's be collapsed into one?

Comment (mcmcgrath13) [reply src/components/dashboard/KeyPerformanceIndicator.vue:3]: `has-text-centered` can probably be moved out to one of these containers instead of repeated

File: src/components/dashboard/KeyPerformanceIndicator.vue [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+<script lang="ts" setup>
+defineProps<{
+  title: string;
+  value: number | string;
+}>();
+</script>
+
+<style scoped>
+.kpi-container {
+  max-width: 200px;
```
Comment (mcmcgrath13) [inline src/components/dashboard/KeyPerformanceIndicator.vue:23]: should probably be in `rem` instead of `px`

File: src/components/dashboard/GapByRace.vue (hunk 8 above)
Comment (eldu) [reply src/components/dashboard/GapByRace.vue:27]: I'll bring back `minVaxRace` as a computed value that is set to `activeStats[0]` for readibility

File: src/components/dashboard/GapByRace.vue (hunk 9 above)
Comment (eldu) [reply src/components/dashboard/GapByRace.vue]: Yes it can be! I've moved it

File: src/components/dashboard/GapByRace.vue (hunk 10 above)
Comment (eldu) [reply src/components/dashboard/GapByRace.vue:161]: I've updated the `minVaxRace` variable

File: src/components/dashboard/KeyPerformanceIndicator.vue (hunk 11 above)
Comment (eldu) [reply src/components/dashboard/KeyPerformanceIndicator.vue]: Yes! I've cleaned up the markup here

File: src/components/dashboard/KeyPerformanceIndicator.vue (hunk 12 above)
Comment (eldu) [reply src/components/dashboard/KeyPerformanceIndicator.vue]: Updated to `20rem`

**Review by maxwellkrieger — APPROVED** [review]
🙌🏼

**Review by s-bessey — APPROVED** [review]

File: src/components/dashboard/GapByRace.vue [hunk 13]
```diff
@@ -58,37 +145,67 @@ const activeStats = computed(() => {
...
-  activeStats.value.forEach((stat) => {
-    if (stat.pct < minRow.pct) {
-      minRow = stat;
+  for (let i = 0; i < activeStats.value.length; i++) {
+    if (!isNaN(activeStats.value[i].pct)) {
+      return activeStats.value[i];
     }
-  });
-  return minRow;
+  }
```
Comment (mcmcgrath13) [inline src/components/dashboard/GapByRace.vue:173]: I'd swap this out for a more modern js loop style (does the same exact thing though)
```js
for (const activeStat of activeStats.value) {
  if (!isNaN(activeStat.pct)) {
    return activeStat;
  }
}
```

File: src/components/dashboard/GapByRace.vue [hunk 14]
```diff
@@ -58,37 +145,67 @@ const activeStats = computed(() => {
...
   place-content: center;
 }
+
+.visible {
+  visibility: visible;
+}
+
+.hidden {
+  visibility: hidden;
+}
```
Comment (mcmcgrath13) [inline src/components/dashboard/GapByRace.vue:198]: there's a bulma utility for this: `is-invisible`

do we need the visible one as well as that's the default value I think?

**Review by mcmcgrath13 — APPROVED** [review]
A couple minor questions - looks great! 🦘

File: src/components/dashboard/GapByRace.vue (hunk 13 above)
Comment (eldu) [reply src/components/dashboard/GapByRace.vue]: Ahh I was psyching myself out which for/or or for/each ran iteratively or in parallel. Good to know that for/or is iterative! Updated !

File: src/components/dashboard/GapByRace.vue (hunk 14 above)
Comment (eldu) [reply src/components/dashboard/GapByRace.vue]: I've updated to use `is-invisible` !
